From b23f0499915fdf0eb03d1639d5b0a54f3f7cf7bd Mon Sep 17 00:00:00 2001
From: Zumeng Chen <zumeng.chen@windriver.com>
Date: Sun, 25 Jul 2010 18:56:52 -0700
Subject: [PATCH] fs/namei/LSB: donot call dget when dentries with zero ref counter

linux/dcache.h has the following comment:

  dget() should never be called for dentries with zero reference counter.
  For these cases(preferably none, functions in dcache.c are sufficient
  for normal needs and they take necessary precautions) you should hold
  dcache_lock and call dget_locked() instead of dget().

So we'll enforce this by locking around dget()

Signed-off-by: Zumeng Chen <zumeng.chen@windriver.com>
---
 fs/namei.c |    4 +++-
 1 files changed, 3 insertions(+), 1 deletions(-)

diff --git a/fs/namei.c b/fs/namei.c
index 2810aef..6c2eeba 100644
--- a/fs/namei.c
+++ b/fs/namei.c
@@ -2149,7 +2149,9 @@ SYSCALL_DEFINE2(mkdir, const char __user *, pathname, int, mode)
  */
 void dentry_unhash(struct dentry *dentry)
 {
-	dget(dentry);
+	spin_lock(&dcache_lock);
+	dget_locked(dentry);
+	spin_unlock(&dcache_lock);
 	shrink_dcache_parent(dentry);
 	spin_lock(&dcache_lock);
 	spin_lock(&dentry->d_lock);
-- 
1.6.5.2

